server/src/main/java/io/druid/query/ResultLevelCachingQueryRunner.java

```diff
@@ -0,0 +1,298 @@
+import io.druid.client.cache.CacheConfig;
+import io.druid.java.util.common.StringUtils;
+import io.druid.java.util.common.guava.Sequence;
+import io.druid.java.util.common.guava.SequenceWrapper;
+import io.druid.java.util.common.guava.Sequences;
+import io.druid.java.util.common.logger.Logger;
+import io.druid.server.QueryResource;
+
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+
+public class ResultLevelCachingQueryRunner<T> implements QueryRunner<T>
+{
+  private static final Logger log = new Logger(ResultLevelCachingQueryRunner.class);
+  private final QueryRunner baseRunner;
+  private final ListeningExecutorService backgroundExecutorService;
+  private QueryToolChest queryToolChest;
+  private ObjectMapper objectMapper;
+  private final Cache cache;
+  private final CacheConfig cacheConfig;
+  private final boolean useResultCache;
+  private final boolean populateResultCache;
+  private final Query<T> query;
+  private final CacheStrategy<T, Object, Query<T>> strategy;
+
+
+  public ResultLevelCachingQueryRunner(
+      QueryRunner baseRunner,
+      QueryToolChest queryToolChest,
+      Query<T> query,
+      ObjectMapper objectMapper,
+      ExecutorService cachingExec,
+      Cache cache,
+      CacheConfig cacheConfig
+  )
+  {
+    this.baseRunner = baseRunner;
+    this.queryToolChest = queryToolChest;
+    this.backgroundExecutorService = MoreExecutors.listeningDecorator(cachingExec);
+    this.objectMapper = objectMapper;
+    this.cache = cache;
+    this.cacheConfig = cacheConfig;
+    this.query = query;
+    this.strategy = queryToolChest.getCacheStrategy(query);
+    this.populateResultCache = ResultLevelCacheUtil.populateResultLevelCacheOnBrokers(query, strategy, cacheConfig);
+    this.useResultCache = ResultLevelCacheUtil.useResultLevelCacheOnBrokers(query, strategy, cacheConfig);
+  }
+
+  @Override
+  public Sequence<T> run(QueryPlus queryPlus, Map responseContext)
+  {
+    if (useResultCache) {
+
+      final String cacheKeyStr = StringUtils.fromUtf8(strategy.computeCacheKey(query));
+      @Nullable
+      final byte[] cachedResultSet = fetchResultsFromResultLevelCache(cacheKeyStr);
+      String existingResultSetId = extractEtagFromResults(cachedResultSet);
+
+      responseContext.put(CacheConfig.ENABLE_RESULTLEVEL_CACHE, true);
+
+      if (existingResultSetId != null) {
+        responseContext.put(QueryResource.EXISTING_RESULT_ID, existingResultSetId);
+      }
+
+      Sequence<T> resultFromClient = baseRunner.run(
+          queryPlus,
+          responseContext
+      );
+      String newResultSetId = (String) responseContext.get(QueryResource.HEADER_ETAG);
+
+      @Nullable
+      ResultLevelCachePopulator resultLevelCachePopulator = createResultLevelCachePopulator(
```

Comment:
this should be created only necessary e.g. not if returned etag matches the one that was already stored in cache and you're not going to populate anything in cache.